File: 07-Repeticion/05-Etiquetas/03-Java/Etiquetas.java
/*
   Programa para verificar si una etiqueta de reciclaje es válida
   y reportar los campos con error.

   Formato: T-aaaa-mm
   - T: tipo de residuo
     (P: plástico, V: vidrio, M: metal, C: cartón/papel, O: orgánico)
   - aaaa: año (> 2025)
   - mm: mes (01-12)

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Septiembre 2025
   Licencia: GNU GPL v3
*/

import modulo.Util;

final int LONGITUD_ETIQUETA = 9;

final char SEPARADOR = '-';
final int POSICION_SEPARADOR1 = 1;
final int POSICION_SEPARADOR2 = 6;

final int POSICION_TIPO = 0;

final int POSICION_ANIO = 2;
final int LONGITUD_ANIO = 4;

final int POSICION_MES = 7;
final int LONGITUD_MES = 2;

final int MINIMO_ANIO = 2026;
final int MINIMO_MES = 1;
final int MAXIMO_MES = 12;

final String TIPOS_VALIDOS = "PVMCO";

final int ETIQUETA_OK    = 0;
final int ERROR_LONGITUD = 1;
final int ERROR_FORMATO  = 2;
final int ERROR_TIPO     = 4;
final int ERROR_ANIO     = 8;
final int ERROR_MES      = 16;

void main() {
	var etiqueta = Util.ingresarTexto("Ingrese la etiqueta a analizar: ");  

	var codigo = verificarEtiquetaReciclaje(etiqueta);

	var reporteEtiqueta = generarReporteEtiqueta(etiqueta, codigo);

	Util.mostrarMensaje(reporteEtiqueta);
}

int verificarEtiquetaReciclaje(String etiqueta) {
	int codigo = verificarLongitud(etiqueta);
	if (codigo != ERROR_LONGITUD) {
		codigo  = verificarFormato(etiqueta)  |
		          verificarTipo(etiqueta)     |
				  		verificarAnio(etiqueta)     |
				  		verificarMes(etiqueta);
	}

	return codigo;
}

int verificarLongitud(String etiqueta) {
	var codigoError = ETIQUETA_OK;

	if (etiqueta.length() != LONGITUD_ETIQUETA ) {
		codigoError = ERROR_LONGITUD;
	}

	return codigoError;
}

int verificarFormato(String etiqueta) {
	var codigoError = ETIQUETA_OK;

	if (etiqueta.charAt(POSICION_SEPARADOR1) != SEPARADOR || 
	    etiqueta.charAt(POSICION_SEPARADOR2) != SEPARADOR) {
		codigoError = ERROR_FORMATO;
	}

	return codigoError;
}

int verificarTipo(String etiqueta) {
	var codigoError = ETIQUETA_OK;

	var tipo = etiqueta.charAt(POSICION_TIPO);

	if (!TIPOS_VALIDOS.contains(Character.toString(tipo)) ) {
		codigoError = ERROR_TIPO;
	}

	return codigoError;
}

int verificarAnio(String etiqueta) {
	var codigoError = ETIQUETA_OK;

	var anioEtiqueta = etiqueta.substring(POSICION_ANIO, POSICION_ANIO + LONGITUD_ANIO);

	if (anioEtiqueta.chars().allMatch(Character::isDigit)) {
		int anio = Integer.parseInt(anioEtiqueta);
		if (anio < MINIMO_ANIO) {
			codigoError = ERROR_ANIO;
		}
	} else {
		codigoError = ERROR_ANIO;
	}

	return codigoError;
}

int verificarMes(String etiqueta) {
	var codigoError = ETIQUETA_OK;

	var mesEtiqueta = etiqueta.substring(POSICION_MES, POSICION_MES + LONGITUD_MES);

	if (mesEtiqueta.chars().allMatch(Character::isDigit)) {
		var mes = Integer.parseInt(mesEtiqueta);
		if (mes < MINIMO_MES || mes > MAXIMO_MES) {
			codigoError =  ERROR_MES;
		}
	} else {
		codigoError =  ERROR_MES;
	}

	return codigoError;
}

String generarReporteEtiqueta(String etiqueta, int codigo) {
	var reporte = String.format("\nLa  etiqueta \"%s\" ", etiqueta );

	if (codigo == ETIQUETA_OK) {
		reporte += "es válida.\n";
	} else {
		reporte += "es inválida por tener:";

		if ((codigo & ERROR_LONGITUD) != 0) { 
			reporte += "\n  - Error en la longitud"; 
		}
		
		if ((codigo & ERROR_FORMATO) != 0)  { 
			reporte += "\n  - Error en el formato";  
		}

		if ((codigo & ERROR_TIPO) != 0)     { 
			reporte += "\n  - Error en el tipo";     
		}

		if ((codigo & ERROR_ANIO) != 0)     { 
			reporte += "\n  - Error en el año";      
		}

		if ((codigo & ERROR_MES) != 0)      { 
			reporte += "\n  - Error en el mes";      
		}
	}

	return reporte + "\n";
}

File: 07-Repeticion/05-Etiquetas/01-Python/Etiquetas.py
"""
    Programa para verificar si una etiqueta de reciclaje es válida 
    y reportar los campos con error.

    Formato: T-aaaa-mm
    - T: tipo de residuo 
      (P: plástico, V: vidrio, M: metal, C: cartón/papel, O: orgánico)
    - aaaa: año (> 2025)
    - mm: mes (01-12)

    Autor: Julián Esteban Gutiérrez Posada
    Fecha: Septiembre 2025
    Licencia: GNU GPL v3
"""

from modulo.util import ingresar_texto, mostrar_mensaje

LONGITUD_ETIQUETA = 9  # T-aaaa-mm

SEPARADOR           = "-"
POSICION_SEPARADOR1 = 1
POSICION_SEPARADOR2 = 6

POSICION_TIPO       = 0

POSICION_ANIO       = 2
LONGITUD_ANIO       = 4

POSICION_MES        = 7
LONGITUD_MES        = 2

MINIMO_ANIO         = 2026
MINIMO_MES          = 1
MAXIMO_MES          = 12

TIPOS_VALIDOS       = "PVMCO"

ETIQUETA_OK         = 0   # 0

ERROR_LONGITUD      = 1   # 2^0
ERROR_FORMATO       = 2   # 2^1
ERROR_TIPO          = 4   # 2^2
ERROR_ANIO          = 8   # 2^3
ERROR_MES           = 16  # 2^4

def main():
    etiqueta =  ingresar_texto("Ingrese la etiqueta a analizar: ")
    
    codigo = verificar_etiqueta_reciclaje(etiqueta)
    
    reporte_etiqueta = generar_reporte_etiqueta(etiqueta, codigo)

    mostrar_mensaje(reporte_etiqueta)

def verificar_etiqueta_reciclaje(etiqueta):
    codigo  = verificar_logitud(etiqueta)
 
    if codigo != ERROR_LONGITUD:
        codigo = verificar_formato(etiqueta) | \
                 verificar_tipo(etiqueta)    | \
                 verificar_anio(etiqueta)    | \
                 verificar_mes(etiqueta)
    
    return codigo

def verificar_logitud(etiqueta):
    codigo_error = ETIQUETA_OK

    if len(etiqueta) != LONGITUD_ETIQUETA:
        codigo_error = ERROR_LONGITUD

    return codigo_error

def verificar_formato(etiqueta):
    codigo_error = ETIQUETA_OK

    if etiqueta[POSICION_SEPARADOR1] != SEPARADOR or \
       etiqueta[POSICION_SEPARADOR2] != SEPARADOR:
        codigo_error = ERROR_FORMATO

    return codigo_error

def verificar_tipo(etiqueta):
    codigo_error = ETIQUETA_OK

    tipo = etiqueta[POSICION_TIPO]

    if tipo not in TIPOS_VALIDOS:
        codigo_error = ERROR_TIPO

    return codigo_error

def verificar_anio(etiqueta):
    codigo_error = ETIQUETA_OK

    anio_etiqueta = etiqueta[POSICION_ANIO:POSICION_ANIO + LONGITUD_ANIO]

    if anio_etiqueta.isdigit():
        anio = int(anio_etiqueta)
        if anio < MINIMO_ANIO:
            codigo_error = ERROR_ANIO
    else:
        codigo_error = ERROR_ANIO

    return codigo_error

def verificar_mes(etiqueta):
    codigo_error = ETIQUETA_OK

    mes_etiqueda = etiqueta[POSICION_MES:POSICION_MES + LONGITUD_MES] 

    if mes_etiqueda.isdigit():
        mes = int(mes_etiqueda)
        if  mes < MINIMO_MES or mes > MAXIMO_MES:
            codigo_error = ERROR_MES
    else:
        codigo_error = ERROR_MES

    return codigo_error

def generar_reporte_etiqueta(etiqueta, codigo):
    reporte =  f"\nLa  etiqueta \"{etiqueta}\" "

    if codigo == ETIQUETA_OK: 
        reporte += "es válida."
    else:
        reporte += "es inválida por tener:"

        if codigo & ERROR_LONGITUD != 0:  
            reporte += "\n  - Error en la longitud"
            
        if codigo & ERROR_FORMATO  != 0:  
            reporte += "\n  - Error en el formato"

        if codigo & ERROR_TIPO     != 0:  
            reporte += "\n  - Error en el tipo"

        if codigo & ERROR_ANIO     != 0:  
            reporte += "\n  - Error en el año"

        if codigo & ERROR_MES      != 0:  
            reporte += "\n  - Error en el mes"

    return reporte +"\n"

main()